### PR TITLE
perf(mla): optimize SM120 sparse MLA with swapped MMA operands

### PR DESCRIPTION
**Pull request by Lemon7-UP**
### Performance summary
Compared with the existing MG prefill kernel, the swapAB(swapped MMA operands) kernel delivers a **1.31–2.16× speedup** across the tested configurations (64/128 heads and 128–8192 query tokens). It achieves up to **1.95×** speedup with the `auto`
scale format and up to **2.16×** with `arbitrary_fp32`.

### Performance
GPU: NVidia RTX PRO5000, SM120
CUDA: 13.0

`python benchmarks/bench_sparse_mla_sm120.py
`(2048/8192 query tokens are not included in benchmark script)

| KV scale format | Heads | Tokens | Before (µs) | After (µs) | Speedup |
|---|---:|---:|---:|---:|---:|
| auto | 64 | 128 | 424.6 | 324.2 | 1.31× |
| auto | 128 | 128 | 682.7 | 420.4 | 1.62× |
| auto | 64 | 512 | 1377.3 | 802.8 | 1.72× |
| auto | 128 | 512 | 2522.8 | 1321.0 | 1.91× |
| auto | 64 | 2048 | 5169.7 | 2945.0 | 1.76× |
| auto | 128 | 2048 | 9784.6 | 5023.6 | 1.95× |
| auto | 64 | 8192 | 20320.3 | 11403.0 | 1.78× |
| auto | 128 | 8192 | 38686.8 | 19896.3 | 1.94× |
| arbitrary_fp32 | 64 | 128 | 553.6 | 356.9 | 1.55× |
| arbitrary_fp32 | 128 | 128 | 912.0 | 494.1 | 1.85× |
| arbitrary_fp32 | 64 | 512 | 1814.1 | 864.8 | 2.10× |
| arbitrary_fp32 | 128 | 512 | 3422.9 | 1601.5 | 2.14× |
| arbitrary_fp32 | 64 | 2048 | 6835.8 | 3264.5 | 2.09× |
| arbitrary_fp32 | 128 | 2048 | 13437.5 | 6220.8 | 2.16× |
| arbitrary_fp32 | 64 | 8192 | 26861.1 | 13077.9 | 2.05× |
| arbitrary_fp32 | 128 | 8192 | 53355.0 | 24661.7 | 2.16× |

### Motivation
SM12x GPUs provide a limited shared-memory capacity of 99 KiB per CTA and do not provide a WGMMA-style instruction that can issue matrix multiplication directly from shared-memory operands. These constraints create significant register-pressure challenges for MLA kernels.

The MLA output dimension is 512. In a conventional FlashAttention-style kernel, warps are partitioned along the query-head dimension. Because SM12x still uses MMA instructions with an fp8 `m16n8k32` shape, each warp would need to hold a `16 × 512` output tile. With FP32 accumulators, this requires 256 accumulator registers per thread, exceeding the practical per-thread register limit and causing extensive register spilling.

Existing FlashInfer sparse MLA kernels avoid this issue by partitioning warps along the KV dimension. However, this design requires:
- Cross-warp synchronization during softmax.
- Writing the probability matrix `P` to shared memory, then synchronizing and reading `P` back before the PV computation.

The overhead is even larger in the FP8 path. The quantization scale of `V` must first be folded into the softmax probability matrix `P`, after which `P` is quantized again. This introduces additional cross-warp synchronization and shared-memory traffic.

### SwapAB design
The new kernel retains FlashAttention-style warp partitioning along the query-head dimension, but swaps the MMA operands:
- `Q` is used as the B operand.
- `K` is used as the A operand.
- `P` is used as the B operand.
- `V` is used as the A operand.

With this layout, each warp independently owns eight attention heads. Softmax and the requantization of `P` are therefore entirely warp-local, eliminating all cross-warp synchronization from these stages.
The MMA B operand has an eight-column output dimension, so each thread only needs 128 FP32 accumulator registers for the final 512-dimensional output. This substantially reduces register pressure and avoids the severe spilling of the conventional layout.

### Kernel organization
The kernel retains double-buffered KV loading and uses a larger `TileKV = 64` to better amortize and hide the softmax overhead. Given the limited shared-memory capacity, `Q` and its quantization scales remain resident in registers. This also reduces shared-memory bandwidth consumption.

The CTA consists of:
- Eight math warps.
- Four I/O warps.
- Eight attention heads per math warp.
- 64 attention heads per CTA.
As a result, the specialized path supports 64 heads with one CTA and 128 heads with two CTAs per query token.

For the PV GEMM:
- Quantized FP8 `P` is stored with the SM120 8-bit transposed `stmatrix` (`STSM_T`) instruction.
- FP8 `V` is loaded with the SM120 8-bit transposed `ldmatrix` (`LDSM_T`) instruction.
- Each warp performs its PV GEMM independently.
- The shared-memory transpose of `P` requires only warp-local synchronization and no cross-warp synchronization.
Overall, the swapAB layout removes cross-warp synchronization and shared-memory round trips from softmax and probability requantization while substantially reducing register pressure.

### feature
- Keep the existing KV-cache layout and all input parameters unchanged.
- Support both UE8M0 block scaling and arbitrary FP32 KV scales.
- Preserve attention-sink and variable `topk_length` behavior.
- Dispatch the new kernel for DSv3.2/GLM sparse MLA prefill with 64 or 128 heads.
- Extend tests and benchmarks to cover the new path.

### ✅ Pre-commit Checks

- [✅  ] I have installed `pre-commit` by running `pip install pre-commit` (or used your preferred method).
- [✅  ] I have installed the hooks with `pre-commit install`.
- [✅  ] I have run the hooks manually with `pre-commit run --all-files` and fixed any reported issues.

## 🧪 Tests

- [✅  ] Tests have been added or updated as needed.
- [✅  ] All tests are passing (`unittest`, etc.).

The existing DSv3.2 prefill test already covers 64/128 heads with the `auto` scale format, including multiple token counts and optional attention sinks.
This PR extends the `arbitrary_fp32` prefill test from `[8, 32]` to `[8, 32, 64, 128]`, covering the new swapAB path for both supported scale formats. Outputs and LSE are validated against the PyTorch reference.

```bash
pytest tests/attention/test_sparse_mla_sm120.py -v
311 passed, 2 warnings 
```

## Reviewer Notes

The main design choice is to swap the MMA operands so that each warp owns eight heads. This reduces the FP32 output accumulators from 256 to 128 registers per thread and removes cross-warp synchronization from softmax and P
requantization.
The new dispatch only affects DSv3.2/GLM prefill with `topk=2048` and 64/128 heads(the kernel design can naturally support any multiple of 64 heads, but the current dispatch specializes 64 and 128 because they are the most common configurations). All other configurations continue using the existing kernels.

## Summary by CodeRabbit

- **New Features**
  - Added optimized prefill support for selected DSv3.2 and GLM-NSA configurations with 64 or 128 heads.
  - Added support for configurable KV scale formats, including arbitrary FP32 scaling.
  - Improved handling of variable token counts, invalid entries, attention sinks, and empty rows.

- **Tests**
  - Expanded coverage for arbitrary-FP32 prefill configurations with 64- and 128-head models.

- **Benchmarks**
  - Added DSv3.2 prefill benchmark sweeps across supported head and token counts.

### CODE REVIEW

**Comment by Lemon7-UP** [comment]
@flashinfer-bot run

**Comment by Lemon7-UP** [comment]
### Additional Performance

GPU: NVidia RTX PRO 6000 Blackwell Server Edition, SM120
CUDA: 13.0

`python benchmarks/bench_sparse_mla_sm120.py` (2048/8192 query tokens are not included in benchmark script)

| KV scale format | Heads | Tokens | Before (µs) | After (µs) | Speedup |
| --- | ---: | ---: | ---: | ---: | ---: |
| auto | 64 | 128 | 308.0 | 203.8 | 1.51× |
| auto | 128 | 128 | 436.2 | 313.2 | 1.39× |
| auto | 64 | 512 | 873.5 | 569.2 | 1.53× |
| auto | 128 | 512 | 1488.9 | 838.7 | 1.78× |
| auto | 64 | 2048 | 3092.4 | 2001.8 | 1.54× |
| auto | 128 | 2048 | 5794.8 | 3055.6 | 1.90× |
| auto | 64 | 8192 | 12073.0 | 7823.5 | 1.54× |
| auto | 128 | 8192 | 22867.2 | 12053.4 | 1.90× |
| arbitrary_fp32 | 64 | 128 | 396.2 | 209.9 | 1.89× |
| arbitrary_fp32 | 128 | 128 | 585.6 | 358.4 | 1.63× |
| arbitrary_fp32 | 64 | 512 | 1162.1 | 591.9 | 1.96× |
| arbitrary_fp32 | 128 | 512 | 2100.1 | 1018.9 | 2.06× |
| arbitrary_fp32 | 64 | 2048 | 4191.0 | 2077.6 | 2.02× |
| arbitrary_fp32 | 128 | 2048 | 8299.4 | 3821.5 | 2.17× |
| arbitrary_fp32 | 64 | 8192 | 16551.9 | 8166.4 | 2.03× |
| arbitrary_fp32 | 128 | 8192 | 32757.7 | 14990.5 | 2.19× |

**Comment by Lemon7-UP** [comment]
Superseded by https://github.com/flashinfer-ai/flashinfer/pull/4802, which carries this swapAB optimization with authorship preserved as commit https://github.com/flashinfer-ai/flashinfer/pull/4751/changes/c6a60a73de2620984e847c1e7576f94e2176f42c (GitHub author: Lemon7-UP) and extends its coverage. Closing this duplicate to keep review focused on the consolidated PR.

**Comment by bkryu** [comment]
Thanks @Lemon7-UP for this PR. 

#4802 has been merged with preserved authorship. We look forward for more contributions from the community!
